plugins/node/instrumentation-undici/test/fetch.test.ts

```diff
@@ -0,0 +1,429 @@
+          };
+        },
+        headersToSpanAttributes: {
+          requestHeaders: ['foo-client', 'x-requested-with'],
+          responseHeaders: ['foo-server'],
+        },
+        applyCustomAttributesOnSpan: (span, req, res) => {
+          span.setAttribute('user.defined.attribute', 'user.defined.value');
+        },
+      });
+
+      // Do some requests
+      const ignoreResponse = await fetch(
+        `${protocol}://${hostname}:${mockServer.port}/ignore/path`
+      );
+      const reqInit = {
+        headers: new Headers({
+          'user-agent': 'custom',
+          'foo-client': 'bar',
+        }),
+      };
+      assert.ok(
+        ignoreResponse.headers.get('propagation-error'),
+        'propagation is not set for ignored requests'
+      );
+
+      const queryResponse = await fetch(
+        `${protocol}://${hostname}:${mockServer.port}/?query=test`,
+        reqInit
+      );
+      assert.ok(
+        queryResponse.headers.get('propagation-error') == null,
+        'propagation is set for instrumented requests'
+      );
+
+      spans = memoryExporter.getFinishedSpans();
+      const span = spans[0];
+      assert.ok(span, 'a span is present');
+      assert.strictEqual(spans.length, 1);
+      assertSpan(span, {
+        hostname: 'localhost',
+        httpStatusCode: queryResponse.status,
+        httpMethod: 'GET',
+        path: '/',
+        query: '?query=test',
+        reqHeaders: reqInit.headers,
+        resHeaders: queryResponse.headers,
+      });
+      assert.strictEqual(
+        span.attributes['http.request.header.foo-client'],
+        'bar',
+        'request headers from fetch options are captured'
+      );
+      assert.strictEqual(
+        span.attributes['http.request.header.x-requested-with'],
+        'undici',
+        'request headers from requestHook are captured'
+      );
+      assert.strictEqual(
+        span.attributes['http.response.header.foo-server'],
+        'bar',
+        'response headers from the server are captured'
+      );
+      assert.strictEqual(
+        span.attributes['test.hook.attribute'],
+        'hook-value',
+        'startSpanHook is called'
+      );
+      assert.strictEqual(
+        span.attributes['user.defined.attribute'],
+        'user.defined.value',
+        'applyCustomAttributesOnSpan is called'
+      );
+    });
+
+    it('should not create spans without parent if required in configuration', async function () {
+      let spans = memoryExporter.getFinishedSpans();
+      assert.strictEqual(spans.length, 0);
+
+      instrumentation.setConfig({
+        enabled: true,
+        requireParentforSpans: true,
+      });
+
+      const fetchUrl = `${protocol}://${hostname}:${mockServer.port}/?query=test`;
+      const response = await fetch(fetchUrl);
+      // TODO: here we're checking the propagation works even if the instrumentation
+      // is not starting any span. Not 100% sure this is the behaviour we want
```

Comment:
AFAIK that's the correct way to do it and is also how we do it in other instrumentations. :slightly_smiling_face: Otherwise it would result in broken traces without manually creating spans.

```suggestion
```